### ISSUE TITLE
Model is not learning over time

### ISSUE THREAD

**Issue by m-michalek** [flairNLP/flair#1096]
Hello guys,

I want to implement active learning with flair. So my plan is to train **150 models** and for each model, I add **5 sentences** and measure the performance using the f1 score. I pick 5 sentences from the train split and 5 sentences from the dev split, the sentences get places in the corresponding train/dev part of the corpus. The sentences to be added are chosen by the score of the uncertainty of a sentence. Each iteration I train a **completely new model** with the before labeled data.

Below you can see my current result.
AL1 = active learning
PL1 = the sentences get chosen randomly

![image](https://user-images.githubusercontent.com/20742170/64867823-6c039000-d63e-11e9-8d84-d35967a5295a.png)

I use the **CoNNL-03** corpus with basic training configurations of learning_rate=0.1,  mini_batch_size=32 and max_epochs=150)

So you can see the performance is not that great. When I look at the data which gets chosen, the sentences often consist out of 1 or a couple of tokens. Sentences that have already been shown to the model in early iterations will be reselected in later iterations.

An example is the "sentence": -DOCSTART-
It appears very often in the dataset because it indicates where a new document starts. I would expect that if I provide my model with an example, it should predict that kind of sentences with high certainty. When I train a fresh model with only some sentences, including "-DOCSTART-" it predicts it with an 'O' tag but only with a certainty of 0.14 which look very low.

So it looks like, the model is not really learning from the seen examples. 

How can it be that the model selects sentences it had already seen and why is it thinking that short sentences are more uncertain than longer sentences?

**Comment by alanakbik** [comment]
Hello @m-michalek interesting project! For this experiment, what word embeddings are you using?

**Comment by m-michalek** [comment]
Hello, @alanakbik thank you!

I'm only using glove.  Training a model even when I pick my sentences randomly can take two days. I experience that when I use more complex or more than one embedding, the training takes much more time.

**Comment by alanakbik** [comment]
Ah yes that explains it. GloVe does not have any handling for out-of-vocabulary words. So if a token like -DOCSTART- comes in, GloVe will assign the same empty vector it gives to all words it does not know. But since this vector is the same for all unknown words, the model does not know that two sentences with -DOCSTART- are the same (from the vector it could also be another unknown word). Also, a sentence consisting only of an out-of-vocabulary word will have high uncertainty which probably explains this. 

Long story short: Try changing to an embedding with handling for OOV words, for instance `BytePairEmbeddings('en')`. Or better do a `StackedEmbeddings` of both `WordEmbeddings('glove')` and `BytePairEmbeddings('en')`.

**Comment by m-michalek** [comment]
Thank you, I will try that out. I hope the training time doesn't add up to much time because of the extra embedding. 

Is there a way to decrease training time? I tried to load the trained model from the predecessor iteration and construct a new custom corpus only consist of the new 5 training and 5 dev sentences. When constructing the corpus I always use an empty list for the test part of the corpus when initializing it to reduce training time. Then I train the already pretrained model with the new corpus.

I experienced that the f score was always much lower than a model with the same data but always trained with 100% of the labeled data. Sometimes I added only the 5 sentences and the f score dropped drastically.

If I could train my models in an incremental way, I would train much faster since the training would change from an exponential problem to a linear one.

Do you have some tips for me here? Since I have to retrain now all of my models because of the wrong embedding it would be really nice to speed up the training process.

The incremental training took 30 minutes with 100 sentences and 10 iterations instead of 4 hours. 150 iterations blow up the process.

**Comment by alanakbik** [comment]
You could try using the current master branch instead of the latest release. The master branch has many significant performance improvements so that could speed things up. 

Another thing to observe is that is your training data is only a few sentences, the default mini_batch_size of 32 might be too high since this means too few update steps per epoch. Maybe this explains the fluctuations in accuracy. For small data set sizes we generally recommend smaller mini batch sizes (and accordingly a smaller learning rate).

**Comment by m-michalek** [comment]
What values would you recommend? mini_batch_size of 16 and learning rate of 0.01?

**Comment by alanakbik** [comment]
Maybe even less that that. Have you found good parameters?

**Comment by tombburnell** [comment]
Hi - Out of interest do you have rough guidelines for choosing learning rates and batch sizes given corpus size?
